pkg/mutators/podtemplatespec.go

```diff
@@ -25,92 +29,118 @@ func NewPodTemplateSpec(opts ...OptionsFunc) *PodTemplateSpec {
 	return pts
 }
 
-func (p *PodTemplateSpec) Mutate(inputPts corev1.PodTemplateSpec, limitRangeMemory *limitrange.Config) (corev1.PodTemplateSpec, error) {
+func (p *PodTemplateSpec) Mutate(ctx context.Context, inputPts corev1.PodTemplateSpec, limitRangeMemory *limitrange.Config) (corev1.PodTemplateSpec, error) {
 
 	pts := *inputPts.DeepCopy()
 	if limitRangeMemory == nil {
 		return pts, fmt.Errorf("invalid limit range config")
 	}
 
-	if err := p.setAndValidateResourceRequirements(pts.Spec.InitContainers, limitRangeMemory); err != nil {
+	if err := p.setAndValidateResourceRequirements(ctx, pts.Spec.InitContainers, limitRangeMemory); err != nil {
 		return pts, err
 	}
 
-	if err := p.setAndValidateResourceRequirements(pts.Spec.Containers, limitRangeMemory); err != nil {
+	if err := p.setAndValidateResourceRequirements(ctx, pts.Spec.Containers, limitRangeMemory); err != nil {
 		return pts, err
 	}
 
 	return pts, nil
 }
 
-func (p *PodTemplateSpec) setAndValidateResourceRequirements(containers []corev1.Container, limitRangeMemory *limitrange.Config) error {
+func (p *PodTemplateSpec) setAndValidateResourceRequirements(ctx context.Context, containers []corev1.Container, limitRangeMemory *limitrange.Config) error {
 	for idx := range containers {
 		container := &containers[idx]
 
-		p.setMemoryRequest(container, limitRangeMemory)
-		p.setMemoryLimit(container, limitRangeMemory)
+		p.setMemoryRequest(ctx, container, limitRangeMemory)
+		p.setMemoryLimit(ctx, container, limitRangeMemory)
```

Comment:
Because memory limit values are a function of request values, setMemoryLimit depends on the container having been actually been mutated for the memory limits to be set properly. In dry-run you will never log limit mutations for replication controllers because the setMemoryLimit will calculate a zero value if the namespace doesn't have a default limit in the limitrange.

Reply:
That's a great catch, thanks!

Changed dry-run to now use a DeepCopy of the container to do the setting and validating. But it won't modify the original (actual) containers passed into Mutate.

Updated test cases to reflect this behavior.